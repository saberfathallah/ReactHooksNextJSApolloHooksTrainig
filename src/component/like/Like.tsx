import React from "react";
import { createStyles, makeStyles } from "@material-ui/core/styles";
import IconButton from "@material-ui/core/IconButton";
import Badge from "@material-ui/core/Badge";
import LikeOn from "@public/assets/images/icons/likeOn.svg";
import LikeOff from "@public/assets/images/icons/likeOff.svg";

const useStyles = makeStyles(() =>
  createStyles({
    heart: {},
  })
);

interface dislikeVariables {
  variables: {
    postId: string;
  };
}

interface addLikeVariables {
  variables: {
    postId: string;
  };
}

interface LikeProps {
  postId: string;
  isLiked: boolean;
  nombreOffLikes: number;
  dislike: (variables: dislikeVariables) => any;
  addLike: (variables: addLikeVariables) => any;
}

const Like: React.FC<LikeProps> = (props: LikeProps) => {
  const { postId, isLiked, nombreOffLikes, dislike, addLike } = props;
  const classes = useStyles();

  return (
    <>
      {isLiked ? (
        <IconButton
          onClick={(): Promise<object> =>
            dislike({
              variables: { postId },
            })
          }
          className={classes.heart}
          aria-label="show 11 new notifications"
          color="inherit"
        >
          <Badge badgeContent={nombreOffLikes}>
            <LikeOn style={{ width: "28px" }} />
          </Badge>
        </IconButton>
      ) : (
        <IconButton
          onClick={(): Promise<object> =>
            addLike({
              variables: { postId },
            })
          }
          className={classes.heart}
          aria-label="show 11 new notifications"
          color="inherit"
        >
          <Badge badgeContent={nombreOffLikes || 0}>
            <LikeOff style={{ width: "28px" }} />
          </Badge>
        </IconButton>
      )}
    </>
  );
};

export default Like;
