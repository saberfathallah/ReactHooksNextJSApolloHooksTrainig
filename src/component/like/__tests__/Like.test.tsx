import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import Like from "../Like";

describe("Like", () => {
  const addLike = jest.fn();
  const dislike = jest.fn();

  const props = {
    dislike: dislike,
    addLike: addLike,
    nombreOffLikes: 1,
    isLiked: true,
    postId: "5e89c6878245cd0d75dec510",
  };

  it("should display dislike snapshot", () => {
    const tree = renderer.create(<Like {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should display like snapshot", () => {
    props.isLiked = false;

    const tree = renderer.create(<Like {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it("should display 0 when nombreOffLikes===null likes snapshot", () => {
    props.isLiked = true;
    props.nombreOffLikes = null;

    const tree = renderer.create(<Like {...props} />);
    expect(tree).toMatchSnapshot();
  });


  it("shouldn't call dislike should call addLike", () => {
    props.isLiked = false;
    props.nombreOffLikes = 1;

    const wrapper = mount(<Like {...props} />);
    wrapper.find("button").simulate("click");
    expect(addLike).toHaveBeenCalledTimes(1);
    expect(dislike).not.toHaveBeenCalled();
  });

  it("shouldn't call addLike should call dislike", () => {
    const addLike = jest.fn();
    props.isLiked = true;
    props.addLike = addLike;

    const wrapper = mount(<Like {...props} />);
    wrapper.find("button").simulate("click");
    expect(dislike).toHaveBeenCalledTimes(1);
    expect(addLike).not.toHaveBeenCalled();
  });
});
