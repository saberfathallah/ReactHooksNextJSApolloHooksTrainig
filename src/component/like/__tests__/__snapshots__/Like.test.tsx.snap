// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Like should display 0 when nombreOffLikes===null likes snapshot 1`] = `
<button
  aria-label="show 11 new notifications"
  className="MuiButtonBase-root MuiIconButton-root makeStyles-heart-1 MuiIconButton-colorInherit"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiIconButton-label"
  >
    <span
      className="MuiBadge-root"
    >
      <svg
        data-file-name="SvgLikeOn"
        style={
          Object {
            "width": "28px",
          }
        }
      />
      <span
        className="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-invisible"
      />
    </span>
  </span>
</button>
`;

exports[`Like should display dislike snapshot 1`] = `
<button
  aria-label="show 11 new notifications"
  className="MuiButtonBase-root MuiIconButton-root makeStyles-heart-1 MuiIconButton-colorInherit"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiIconButton-label"
  >
    <span
      className="MuiBadge-root"
    >
      <svg
        data-file-name="SvgLikeOn"
        style={
          Object {
            "width": "28px",
          }
        }
      />
      <span
        className="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle"
      >
        1
      </span>
    </span>
  </span>
</button>
`;

exports[`Like should display like snapshot 1`] = `
<button
  aria-label="show 11 new notifications"
  className="MuiButtonBase-root MuiIconButton-root makeStyles-heart-1 MuiIconButton-colorInherit"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiIconButton-label"
  >
    <span
      className="MuiBadge-root"
    >
      <svg
        data-file-name="SvgLikeOff"
        style={
          Object {
            "width": "28px",
          }
        }
      />
      <span
        className="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle"
      >
        1
      </span>
    </span>
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;
